core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java

```diff
@@ -809,14 +810,14 @@ private static String toSql(RelNode root, SqlDialect dialect,
         + " COUNT(*) AS \"C\"\n"
         + "FROM \"foodmart\".\"product\"\n"
         + "GROUP BY ROLLUP(\"product_class_id\", \"brand_name\")\n"
-        + "ORDER BY \"product_class_id\", \"brand_name\", COUNT(*)";
+        + "ORDER BY \"product_class_id\", \"brand_name\", 3";
     final String expectedMysql = "SELECT `product_class_id`, `brand_name`,"
         + " COUNT(*) AS `C`\n"
         + "FROM `foodmart`.`product`\n"
         + "GROUP BY `product_class_id`, `brand_name` WITH ROLLUP\n"
         + "ORDER BY `product_class_id` IS NULL, `product_class_id`,"
         + " `brand_name` IS NULL, `brand_name`,"
-        + " COUNT(*) IS NULL, COUNT(*)";
+        + " 3 IS NULL, 3";
```

Comment:
Is `order by 3 is null` valid here? In [Calcite](https://github.com/apache/calcite/blob/56e9deff51942fe7673096aaf5f50544f47f1411/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java#L6677-L6680), we'll take this not a 'field reference'. I'm not sure how other databases behaves in this case.

Reply:
I think it valid here.
I try to make a test to compare the relNode compare '3 is null' sql with the original SQL and the 'count(*) is null' sql.
Using default SqlParser config.

1. For original sql:
```
select "product_class_id", "brand_name",count(*) as c
from "product"
group by rollup("product_class_id", "brand_name")
order by 1, 2, 3
```
relNode is
```
LogicalSort(sort0=[$0], sort1=[$1], sort2=[$2], dir0=[ASC], dir1=[ASC], dir2=[ASC])
  LogicalAggregate(group=[{0, 1}], groups=[[{0, 1}, {0}, {}]], C=[COUNT()])
    LogicalProject(product_class_id=[$0], brand_name=[$2])
      JdbcTableScan(table=[[foodmart, product]])
```


2. For 'count(*) is null' sql:
```
select "product_class_id", "brand_name", count(*) as c
from "product"
group by rollup("product_class_id", "brand_name")
order by "product_class_id", "brand_name", count(*) is null, count(*)
```
relNode is
```
LogicalProject(product_class_id=[$0], brand_name=[$1], C=[$2])
  LogicalSort(sort0=[$0], sort1=[$1], sort2=[$3], sort3=[$2], dir0=[ASC], dir1=[ASC], dir2=[ASC], dir3=[ASC])
    LogicalProject(product_class_id=[$0], brand_name=[$1], C=[$2], EXPR$3=[false])
      LogicalAggregate(group=[{0, 1}], groups=[[{0, 1}, {0}, {}]], C=[COUNT()])
        LogicalProject(product_class_id=[$0], brand_name=[$2])
          JdbcTableScan(table=[[foodmart, product]])
```


3. For '3 is null' sql:
```
select "product_class_id", "brand_name", count(*) as c
from "product"
group by rollup("product_class_id", "brand_name")
order by 1, 2, 3 is null, 3
```
relNode is
```
LogicalProject(product_class_id=[$0], brand_name=[$1], C=[$2])
  LogicalSort(sort0=[$0], sort1=[$1], sort2=[$3], sort3=[$2], dir0=[ASC], dir1=[ASC], dir2=[ASC], dir3=[ASC])
    LogicalProject(product_class_id=[$0], brand_name=[$1], C=[$2], EXPR$3=[false])
      LogicalAggregate(group=[{0, 1}], groups=[[{0, 1}, {0}, {}]], C=[COUNT()])
        LogicalProject(product_class_id=[$0], brand_name=[$2])
          JdbcTableScan(table=[[foodmart, product]])
```

The relNode of '3 is null' sql and 'count(*) is null' sql is almost some. Is my understanding correct?
So I think it's fine here. And I test '3 is null' sql in MySQL, it works.

Reply:
No, what I mean it that we only takes pure literal as a field reference in Calcite, if it's an expression, it won't be taken as a field reference. Your second case shows what I described above.

To be more precise, `3 is null` equals `false`, not `count(*) is null` in this case.

I'm not saying Calcite is definitely correct for current behavior. We need to investigate what other databases behave when the literal is in a expression, and let Calcite be consistent with the majority of them.

If we prefer to not take literal in expressions as a field reference, then your change now would make it incorrect.

Reply:
I got it. We shouldn't take literal in experssion as field reference.
How about keep original expression when we need to sort null values in order by?
For example, the output sql will be 
```
select "product_class_id", "brand_name", count(*) as c
from "product"
group by rollup("product_class_id", "brand_name")
order by 1, 2, count(*) is null, 3
```

Reply:
This sounds good to me.

Reply:
I fixed it. Please review again, thanks.